sql/lease.go

```diff
@@ -0,0 +1,500 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/cockroachdb/cockroach/util/retry"
+	"github.com/gogo/protobuf/proto"
+)
+
+// TODO(pmattis): Periodically renew leases for tables that were used recently and
+// for which the lease will expire soon.
+
+const (
+	leaseDuration    = int64(5 * time.Minute)
+	minLeaseDuration = int64(time.Minute)
+)
+
+var (
+	errLeaseVersionChanged = errors.New("lease version changed")
+)
+
+func nanosToTime(nanos int64) time.Time {
+	return time.Unix(0, nanos)
+}
+
+func nanosToDTimestamp(nanos int64) parser.DTimestamp {
+	return parser.DTimestamp{Time: nanosToTime(nanos)}
+}
+
+// LeaseState holds the state for a lease. Exported only for testing.
+type LeaseState struct {
+	TableDescriptor
+	expiration int64
+	refcount   int
+}
+
+func (s *LeaseState) String() string {
+	return fmt.Sprintf("%d:%d", s.Version, s.expiration)
+}
+
+// Expiration returns the expiration time of the lease.
+func (s *LeaseState) Expiration() time.Time {
+	return nanosToTime(s.expiration)
+}
+
+// Refcount returns the reference count of the lease.
+func (s *LeaseState) Refcount() int {
+	return s.refcount
+}
+
+// LeaseStore implements the operations for acquiring and releasing leases and
+// publishing a new version of a descriptor. Exported only for testing.
+type LeaseStore struct {
+	db     client.DB
+	clock  *hlc.Clock
+	nodeID uint32
+}
+
+// jitteredLeaseDuration returns a randomly jittered duration from the interval
+// [0.75 * leaseDuration, 1.25 * leaseDuration].
```

Comment:
a smaller 5%? jitter is plenty. The time is in minutes and 5% allows you think in terms of approx 5 minutes rather than worrying about the possibility of < 4 minute expirations.